Avoid constructing InputType in HTMLInputElement::updateType whenever possible

rniwa opened the pull request:
#### 62339dafa6fc921bb8831a2a1edaaa437c4831cd
<pre>
Avoid constructing InputType in HTMLInputElement::updateType whenever possible
<a href="https://bugs.webkit.org/show_bug.cgi?id=245042">https://bugs.webkit.org/show_bug.cgi?id=245042</a>

Reviewed by Cameron McCormack and Chris Dumez.

Avoid constructing a new InputType just to check whether the type has changed or not.

* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::updateType):
(WebCore::HTMLInputElement::initializeInputType):
* Source/WebCore/html/InputType.cpp:
(WebCore::createInputTypeFactoryMap):
(WebCore::findFactory):
(WebCore::InputType::createIfDifferent): Renamed from create. Optionally takes the current InputType.
* Source/WebCore/html/InputType.h:

Canonical link: <a href="https://commits.webkit.org/254383@main">https://commits.webkit.org/254383@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0b3dbe458b339ac82a434af3dffe7128a5224544

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/88863 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/33428 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/19744 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/98065 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/154546 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/92857 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/31934 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/27537 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/81107 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/92685 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/94492 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/25344 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/75846 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/25297 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/80216 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/80252 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/68261 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/29729 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/14270 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/29459 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/15272 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3074 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/32898 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/38207 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/31586 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/34377 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

heycam commented on Source/WebCore/html/InputType.cpp:
```diff
@@ -91,7 +91,14 @@ using namespace HTMLNames;
+    InputTypeFactoryFunction factoryFunction;
+};
+
+typedef MemoryCompactLookupOnlyRobinHoodHashMap<AtomString, InputTypeFactory> InputTypeFactoryMap;
```
I don't think we need to store the type `AtomString` in both the key and the value. Can we use `MemoryCompactLookupOnlyRobinHoodHashTable<InputTypeFactor>` instead to avoid this? Or, change `findFactory` to return the `MemoryCompactLookupOnlyRobinHoodHashMap::iterator` (or return a pair of the `AtomString` and `InputTypeFactory` pointers?) so that its caller can look at both the key and value?

rniwa replied:
ok, I'd make this function return `std::pair<const AtomString&, InputTypeFactory*>`.

heycam commented on Source/WebCore/html/InputType.cpp:
```diff
@@ -142,27 +149,38 @@ static InputTypeFactoryMap createInputTypeFactoryMap()
 }
 
-Ref<InputType> InputType::create(HTMLInputElement& element, const AtomString& typeName)
+Ref<InputType> InputType::createIfDifferent(HTMLInputElement& element, const AtomString& typeName, InputType* currentInputType)
```
I find it a little bit strange that this function called `createIfDifferent` can return the `currentInputType` object that was passed in, which we then assign back to the same variable we got it from. We end up bumping the refcount up and down in that case, too.

What if instead we make this function return `RefPtr<InputType>`, return `nullptr` if the type already matches, and have `HTMLInputElement::updateType` do:

```c++
auto newType = InputType::createIfDifferent(...);
if (!newType)
    return;
```

rniwa replied:
Sure, that was an alternative design I was thinking of anyway. So done.

heycam reviewed (approved):
r=me with some changes to consider.

weinig commented on Source/WebCore/html/InputType.cpp:
```diff
@@ -143,26 +149,35 @@ static InputTypeFactoryMap createInputTypeFactoryMap()
-    if (UNLIKELY(!factory.second))
-        factory = factoryMap.get().get(typeName.convertToASCIILowercase());
-    return factory;
+    auto& map = factoryMap.get();
```
Not a change you should make, but I am curious where the cutoff for a map like this is where HashMap<AtomString, ...> starts winning over SortedArrayMap.  Time for me to write a microbenchmark.

cdumez replied:
For what it's worth, I tried using a SortedArrayMap to replace the HashMap in this function a while back. Sadly it was a small regression on Speedometer.

cdumez reviewed (approved):
LGTM

rniwa committed:
Avoid constructing InputType in HTMLInputElement::updateType whenever possible
https://bugs.webkit.org/show_bug.cgi?id=245042

Reviewed by Cameron McCormack and Chris Dumez.

Avoid constructing a new InputType just to check whether the type has changed or not.

* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::updateType):
(WebCore::HTMLInputElement::initializeInputType):
* Source/WebCore/html/InputType.cpp:
(WebCore::createInputTypeFactoryMap):
(WebCore::findFactory):
(WebCore::InputType::createIfDifferent): Renamed from create. Optionally takes the current InputType.
* Source/WebCore/html/InputType.h:

Canonical link: https://commits.webkit.org/254383@main
